how-to-guides/12-param-schedulers.ipynb

```diff
@@ -0,0 +1,835 @@
+        " "
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 1,
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "import ast\n",
+        "\n",
+        "import ipywidgets as widgets\n",
+        "import matplotlib.pyplot as plt\n",
+        "import numpy as np"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 2,
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "import torch\n",
+        "from torch.optim import SGD"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "# `PyTorch` schedulers\n",
+        "\n",
+        "We use the `LRScheduler` tool of `PyTorch-Ignite` which allows to easily integrate the schedulers of `PyTorch` and in particular to display the values. "
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 3,
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "from ignite.handlers import LRScheduler"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "> Beware that not all `PyTorch` schedulers are compatible because they have more or less exotic modes."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 22,
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "from torch.optim.lr_scheduler import (\n",
+        "    CosineAnnealingLR,\n",
+        "    CyclicLR,\n",
+        "    ExponentialLR,\n",
+        "    LambdaLR,\n",
+        "    StepLR,\n",
+        "    ExponentialLR,\n",
+        "    MultiplicativeLR,\n",
+        "    MultiStepLR,\n",
+        "    OneCycleLR,\n",
+        "    LambdaLR\n",
+        "    # ReduceLROnPlateau : not compatible\n",
+        "    # CosineAnnealingWarmRestarts : not compatible\n",
+        ")"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 5,
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "def print_optimizer(cls, num_events=50, lr=1.0, *args, **kwargs):\n",
+        "    optimizer = SGD(params=[torch.zeros(10)], lr=lr)\n",
+        "    lr_scheduler = cls(optimizer=optimizer, *args, **kwargs)\n",
+        "    LRScheduler.plot_values(num_events=num_events, lr_scheduler=lr_scheduler)\n",
+        "    plt.title(\"{}\".format(cls))\n",
+        "    plt.show()"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "## LambdaLR\n",
+        "\n",
+        "Sets the learning rate of each parameter group to the initial lr times a given function. When last_epoch=-1, sets initial lr as lr.\n",
+        "\n",
+        "$$\n",
+        "l r_{\\text {epoch}} = l r_{\\text {initial}} * Lambda(epoch)\n",
```

Comment:
lr is modified every step, not epoch. In the PyTorch doc, `optimizer.step()` is called at the epoch end, it’s more flexible in ignite. The formula should not be written using epoch as index name.

Reply:
Oh, my bad. I'm sorry. I didn't think in this manner.